ts/tracer.ts

```diff
@@ -0,0 +1,99 @@
+import { Message } from './message'
+import { GossipsubIWantFollowupTime } from './constants'
+
+/**
+ * IWantTracer is an internal tracer that tracks IWANT requests in order to penalize
+ * peers who don't follow up on IWANT requests after an IHAVE advertisement.
+ * The tracking of promises is probabilistic to avoid using too much memory.
+ *
+ * Note: Do not confuse these 'promises' with JS Promise objects.
+ * These 'promises' are merely expectations of a peer's behavior.
+ */
+export class IWantTracer {
+  getMsgId: (msg: Message) => string
+  /**
+   * Promises to deliver a message
+   * Map per message id, per peer, promise expiration time
+   */
+  promises: Map<string, Map<string, number>>
+  constructor (getMsgId: (msg: Message) => string) {
+    this.getMsgId = getMsgId
+    this.promises = new Map()
+  }
+
+  /**
+   * Track a promise to deliver a message from a list of msgIDs we are requesting
+   * @param {string} p peer id
+   * @param {string[]} msgIds
+   * @returns {void}
+   */
+  addPromise (p: string, msgIds: string[]): void {
+    // pick msgId randomly from the list
+    const ix = Math.floor(Math.random() * msgIds.length)
+    const msgId = msgIds[ix]
+
+    let peers = this.promises.get(msgId)
+    if (!peers) {
+      peers = new Map()
+      this.promises.set(msgId, peers)
+    }
+
+    if (!peers.has(p)) {
+      peers.set(p, Date.now() + GossipsubIWantFollowupTime)
+    }
+  }
+
+  /**
+   * Returns the number of broken promises for each peer who didn't follow up on an IWANT request.
+   * @returns {Map<string, number>}
+   */
+  getBrokenPromises (): Map<string, number> {
+    const now = Date.now()
+    const result = new Map<string, number>()
+
+    this.promises.forEach((peers, msgId) => {
+      peers.forEach((expire, p) => {
+        // the promise has been broken
+        if (expire < now) {
+          // add 1 to result
+          result.set(p, (result.get(p) || 0) + 1)
+          // delete from tracked promises
+          peers.delete(p)
+        }
+      })
+      // clean up empty promises for a msgId
+      if (!peers.size) {
+        this.promises.delete(msgId)
+      }
+    })
+
+    return result
+  }
+
+  /**
+   * Someone delivered a message, stop tracking promises for it
+   * @param {string} p peer id
```

Comment:
It seems that we are not using the peerID

Reply:
The idea is to have the same function signature as `PeerScore#deliverMessage`.
So that we might in the future abstract the different 'tracers' as in go-libp2p-pubsub
See https://github.com/libp2p/go-libp2p-pubsub/blob/master/pubsub.go#L972
Eg: they don't call each one individually:
```
this.score.deliverMessage(p, msg)
this.gossipTracer.deliverMessage(p, msg)
this.tagTracer.deliverMessage(p, msg)
```
They have some sort of registration for each of their tracers, then all tracers get called on a single call to `deliverMessage`.
Note the go-libp2p-pubsub tracer functionality is happening on the pubsub layer, not the gossipsub layer. Aggregating our "tracers" together would be a first step towards moving in that direction. Not saying thats what we _should_ do, but it is a prerequisite.

Reply:
ok, sounds good